File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionRandom.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;

import java.math.BigDecimal;
import java.util.Map;
import java.util.Random;

public class TransformerFunctionRandom extends TransformerFunction {
    final Random random = new Random();

    public TransformerFunctionRandom() {
        super(FunctionDescription.of(
            Map.of(
            "min", ArgumentType.of(ArgType.Number).position(0).defaultValue(BigDecimal.ZERO),
            "max", ArgumentType.of(ArgType.Number).position(1).defaultValue(BigDecimal.ONE)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        BigDecimal min;
        BigDecimal max;
        if (context.has("min")) {
            min = context.getBigDecimal("min");
            max = context.getBigDecimal("max");
        } else {
            var arr = context.getJsonArray(null);
            var adapter = context.getAdapter();
            var length = adapter.size(arr);
            min = length > 0 ? adapter.getNumberAsBigDecimal(adapter.get(arr, 0)) : null;
            max = length > 1 ? adapter.getNumberAsBigDecimal(adapter.get(arr, 1)) : null;
        }
        // sanity check
        if (min == null || max == null || max.compareTo(min) < 0) {
            return null;
        }
        var rand = BigDecimal.valueOf(random.nextDouble());
        var diff = max.subtract(min);
        var result = min.add(diff.multiply(rand, FunctionHelpers.DEFAULT_MATH_CONTEXT));
        // cap scale at max
        if (result.scale() > FunctionHelpers.MAX_SCALE) {
            result = result.setScale(FunctionHelpers.MAX_SCALE, FunctionHelpers.MAX_SCALE_ROUNDING);
        }
        return result;
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionRandom.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { BigDecimal, BigDecimal_ONE, BigDecimal_ZERO, MAX_SCALE, MAX_SCALE_ROUNDING } from "./common/FunctionHelpers";
import BigNumber from "bignumber.js";
import { unwrapNumber } from "../JsonHelpers";

class TransformerFunctionRandom extends TransformerFunction {
  constructor() {
    super({
      allowsArgumentsAsInput: true,
      argsSet: [
        { name: "min", type: ArgType.Number, defaultValue: 0 },
        { name: "max", type: ArgType.Number, defaultValue: 1 },
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    let min: BigNumber | null;
    let max: BigNumber | null;
    if (context.has("min")) {
      min = (await context.getBigDecimal("min")) ?? BigDecimal_ZERO;
      max = (await context.getBigDecimal("max")) ?? BigDecimal_ONE;
    } else {
      const arr = await context.getJsonArray(null);
      if (!arr) {
        return null;
      }
      min = arr[0] ?? BigDecimal_ZERO;
      max = arr[1] ?? BigDecimal_ONE;
    }
    if (!BigDecimal.isBigNumber(min) && min !== null) min = BigDecimal(min);
    if (!BigDecimal.isBigNumber(max) && max !== null) max = BigDecimal(max);
    // sanity check
    if (min === null || max === null || max.lt(min)) {
      return null;
    }
    const rand = new BigDecimal(Math.random());
    const diff = max.minus(min);
    let result = min.plus(diff.multipliedBy(rand));
    // cap scale at max
    if ((result.decimalPlaces() ?? 0) > MAX_SCALE) {
      result = result.decimalPlaces(MAX_SCALE, MAX_SCALE_ROUNDING);
    }
    return unwrapNumber(result);
  }
}

export default TransformerFunctionRandom;
